File: backend-go/internal/application/balance.go
package application

import (
	"time"

	"github.com/google/uuid"

	"github.com/pharaujo/finance/backend-go/internal/domain"
)

// TransactionSlice is the minimum a transaction contributes to any balance
// calculation, mirroring FinanceTracker's record of the same name.
type TransactionSlice struct {
	AccountID         uuid.UUID
	TransferAccountID *uuid.UUID
	CategoryID        *uuid.UUID
	Type              domain.TransactionType
	Amount            domain.Money
	Date              time.Time
}

// DeltaFor is the signed effect of one slice on one specific account. A
// transfer credits its destination and debits its source; income credits, and
// everything else debits.
func DeltaFor(accountID uuid.UUID, slice TransactionSlice) domain.Money {
	if slice.Type == domain.TransactionTransfer &&
		slice.TransferAccountID != nil && *slice.TransferAccountID == accountID {
		return slice.Amount
	}

	if slice.AccountID != accountID {
		return domain.Zero()
	}

	if slice.Type == domain.TransactionIncome {
		return slice.Amount
	}
	return slice.Amount.Neg()
}

// NetWorthDelta is the signed effect on total net worth: transfers between the
// user's own accounts cancel out and contribute nothing.
func NetWorthDelta(slice TransactionSlice) domain.Money {
	switch slice.Type {
	case domain.TransactionIncome:
		return slice.Amount
	case domain.TransactionExpense:
		return slice.Amount.Neg()
	case domain.TransactionTransfer:
		return domain.Zero()
	default:
		return domain.Zero()
	}
}

// BalanceOf is the account's opening balance plus the signed sum of every
// slice, which is how both APIs derive a live balance.
func BalanceOf(account domain.Account, slices []TransactionSlice) domain.Money {
	total := domain.Zero()
	for _, slice := range slices {
		total = total.Add(DeltaFor(account.Id, slice))
	}
	return account.InitialBalance.Add(total)
}

File: backend-rb/lib/services/balance.rb
# frozen_string_literal: true

module Services
  # How a transaction moves an account balance and the net worth.
  #
  # Amounts are always stored positive; the type carries the sign. A transfer
  # debits its source and credits its destination, which is why the destination
  # check comes first -- the destination is not the row's "AccountId".
  module Balance
    module_function

    # What one transaction does to one account's balance.
    def delta_for(account_id, item)
      # The receiving side of a transfer, which the row records as the
      # *transfer* account rather than the owning one.
      if item.type.is?(Domain::Enums::TRANSACTION_TRANSFER) &&
         !item.transfer_account_id.nil? && item.transfer_account_id == account_id
        return item.amount
      end

      return Domain::MONEY_ZERO unless item.account_id == account_id
      return item.amount if item.type.is?(Domain::Enums::TRANSACTION_INCOME)

      # An expense and the paying side of a transfer both debit.
      item.amount.negate
    end

    # What one transaction does to the total. A transfer moves money, so zero.
    def net_worth_delta(item)
      return item.amount if item.type.is?(Domain::Enums::TRANSACTION_INCOME)
      return item.amount.negate if item.type.is?(Domain::Enums::TRANSACTION_EXPENSE)

      Domain::MONEY_ZERO
    end

    # The opening balance plus every movement. Deliberately not rounded.
    def balance_of(account, slices)
      total = slices.reduce(Domain::MONEY_ZERO) { |sum, item| sum.add(delta_for(account.id, item)) }
      account.initial_balance.add(total)
    end
  end
end
